src/bar.rs

```diff
@@ -0,0 +1,205 @@
+            }
+        }
+    }
+}
+
+impl From<ConfigSpaceAccessError> for Error {
+    fn from(e: ConfigSpaceAccessError) -> Self {
+        Error::Access(e)
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Enum representing the "prefetchable" state masks of BARs (bit 3 in the register).
+pub enum PciBarPrefetchable {
+    /// Non-prefetchable BARs have bit 3 set to 0.
+    NotPrefetchable = 0,
+    /// Prefetchable BARs have bit 3 set to 1.
+    Prefetchable = 0x08,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+/// PCI BAR memory region abstraction.
+pub enum PciBarRegion {
+    /// Represents a 32-bit IO space region.
+    Io {
+        /// Start address of the 32-bit IO space region.
+        addr: u32,
+        /// Length of the region, in bytes.
+        len: u32,
+    },
+    /// Represents a 32-bit memory mapped region.
+    Memory32Bit {
+        /// Start address of the 32-bit memory mapped region.
+        addr: u32,
+        /// Length of the region, in bytes.
+        len: u32,
+    },
+    /// Represents a 64-bit memory mapped region.
+    Memory64Bit {
+        /// Start address of the 64-bit memory mapped region.
+        addr: u64,
+        /// Length of the region, in bytes.
+        len: u64,
+    },
+}
+
+impl PciBarRegion {
+    /// Describe an IO space BAR region.
+    pub fn new_io_region(addr: u32, len: u32) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Io { addr, len })
+    }
+
+    /// Describe a memory mapped 32bit BAR region.
+    pub fn new_32bit_mem_region(addr: u32, len: u32) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Memory32Bit { addr, len })
+    }
+
+    /// Describe a memory mapped 64bit BAR region.
+    pub fn new_64bit_mem_region(addr: u64, len: u64) -> Result<Self> {
+        addr.checked_add(len)
+            .ok_or(Error::BarAddressInvalid(addr.into(), len.into()))?;
+        validate_dword_alignment(addr as usize)?;
+        Ok(PciBarRegion::Memory64Bit { addr, len })
+    }
+}
+
+#[derive(Copy, Clone)]
+/// Configuration of a Base Address Register.
+pub struct PciBarConfig {
+    /// Index of the BAR.
+    pub index: usize,
+    /// Region referenced by BAR.
+    pub region: PciBarRegion,
+    /// Prefetchable status.
+    pub prefetchable: PciBarPrefetchable,
+}
+
+#[derive(Copy, Clone)]
+/// Configuration of a ROM Base Address Register.
+pub struct PciRomBarConfig {
+    /// Region referenced by BAR.
+    pub region: PciBarRegion,
+    /// Enable bit.
+    pub enable: bool,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_io_bar_region() {
+        let r = PciBarRegion::new_io_region(0x1000, 4096).unwrap();
+        assert!(
```

Comment:
This should be written with `assert_eq` instead. Same for all the asserts in this file.